File: Trees (BT + BST)/python/preOrderOfTree.py
# Problem Statement: Given the root of a binary tree, return the preorder traversal of its nodes' values.

# Problem Link: https://www.codingninjas.com/studio/problems/preorder-binary-tree_5948?utm_source=striver&utm_medium=website&utm_campaign=a_zcoursetuf

class BinaryTreeNode:
    def __init__(self, data=None):
        self.data = data
        self.left = None
        self.right = None

def pre_order(root):
    if root is None:
        return
        
    print(root.data, end=" ")
    pre_order(root.left)
    pre_order(root.right)

File: Trees (BT + BST)/cpp/preOrderOfTree.cpp
// Problem Statement: Given the root of a binary tree, return the preorder traversal of its nodes' values.

// Problem Link: https://www.codingninjas.com/studio/problems/preorder-binary-tree_5948?utm_source=striver&utm_medium=website&utm_campaign=a_zcoursetuf

#include <iostream>
using namespace std;

template <typename T>
class BinaryTreeNode {
public:
    T data;
    BinaryTreeNode<T>* left;
    BinaryTreeNode<T>* right;

    BinaryTreeNode(T data) {
        this->data = data;
        this->left = NULL;
        this->right = NULL;
    }
};

void preOrder(BinaryTreeNode<int>* root) {
    if (root == NULL) {
        return;
    }
    
    cout << root->data << " ";
    preOrder(root->left);
    preOrder(root->right);
}
